lib/Dialect/SV/Transforms/HWRaiseInOutPorts.cpp

```diff
@@ -0,0 +1,195 @@
+  PortInfo readPort, writePort;
+};
+
+HWInOutPortConversion::HWInOutPortConversion(PortConverterImpl &converter,
+                                             hw::PortInfo port)
+    : PortConversion(converter, port) {
+  // Gather readers and writers (how to handle sv.passign?)
+  for (auto *user : body->getArgument(port.argNum).getUsers()) {
+    if (auto read = dyn_cast<sv::ReadInOutOp>(user))
+      readers.push_back(read);
+    else if (auto write = dyn_cast<sv::AssignOp>(user))
+      writers.push_back(write);
+    else
+      user->emitWarning() << "Use of inout port " << port.name
+                          << " is not supported";
+  }
+
+  if (writers.size() > 1)
+    converter.getModule()->emitWarning()
+        << "Multiple writers of inout port " << port.name
+        << " detected. Will only create an output for the first write";
+}
+
+void HWInOutPortConversion::buildInputSignals() {
+  if (hasReaders()) {
+    Value readValue =
+        converter.createNewInput(origPort, "_rd", origPort.type, readPort);
+
+    // Replace all sv::ReadInOutOp's with the new input.
+    Value origInput = body->getArgument(origPort.argNum);
+    for (auto *user : llvm::make_early_inc_range(origInput.getUsers())) {
+      sv::ReadInOutOp read = dyn_cast<sv::ReadInOutOp>(user);
+      if (!read)
+        continue;
+
+      read.replaceAllUsesWith(readValue);
+      read.erase();
+    }
+  }
+
+  if (hasWriters()) {
+    sv::AssignOp write = writers.front();
+    converter.createNewOutput(origPort, "_wr", origPort.type, write.getSrc(),
+                              writePort);
+    write.erase();
+  }
+}
+
+void HWInOutPortConversion::buildOutputSignals() {
+  // TODO: could support hw.inout outputs (always create read/write ports) -
+  // don't need it for now, though.
+  assert(false && "hw.inout outputs not yet supported");
+}
+
+void HWInOutPortConversion::mapInputSignals(OpBuilder &b, Operation *inst,
+                                            Value instValue,
+                                            SmallVectorImpl<Value> &newOperands,
+                                            ArrayRef<Backedge> newResults) {
+
+  if (hasReaders()) {
+    // Create a read_inout op at the instantiation point. This effectively
+    // pushes the read_inout op from the module to the instantiation site.
+    newOperands[readPort.argNum] =
+        b.create<ReadInOutOp>(inst->getLoc(), instValue).getResult();
+  }
+
+  if (hasWriters()) {
+    // Create a sv::AssignOp at the instantiation point. This effectively
+    // pushes the write op from the module to the instantiation site.
+    Value writeFromInsideMod = newResults[writePort.argNum];
+    b.create<sv::AssignOp>(inst->getLoc(), instValue, writeFromInsideMod);
+  }
+}
+
+void HWInOutPortConversion::mapOutputSignals(
+    OpBuilder &b, Operation *inst, Value instValue,
+    SmallVectorImpl<Value> &newOperands, ArrayRef<Backedge> newResults) {
+  llvm_unreachable("hw.inout outputs not yet supported");
```

Comment:
Above you use an assert whereas here you use this. What should really be used? I think assert is more appropriate.

Reply:
See below - will keep the assert since i nevertheless have to have _something_ in the implementation.